File: sdk/typescript/src/client/signing.ts
/**
 * SDK-side request signing (scheme A, opt-in) — HMAC over a canonical string
 * keyed by the tenant's api_key, with a timestamp + nonce to defend against
 * replay. Must be BYTE-FOR-BYTE identical to the server
 * (`io.github.pinpols.batch.common.security.RequestSignatures`) and the Java
 * SDK signer (`...sdk.internal.RequestSigner`) — they are pinned together by
 * the conformance golden vectors:
 *
 * ```
 *   canonical = UPPER(method) "\n" path "\n" timestamp "\n" nonce "\n" hex(sha256(body))
 *   signature = hex(hmacSha256(apiKey, canonical))     // lowercase hex
 * ```
 *
 * The server defaults the check OFF (`batch.request-signing.enabled=false`);
 * the SDK only attaches the headers when explicitly opted in AND an api_key is
 * present (no key → nothing to sign with).
 */

import { createHash, createHmac, randomUUID } from "node:crypto";

/** Header names — must match the server-side filter (case-insensitive on the wire). */
export const SIGNATURE_HEADERS = {
  timestamp: "X-Batch-Timestamp",
  nonce: "X-Batch-Nonce",
  signature: "X-Batch-Signature",
} as const;

/** hex(sha256(body)) — empty body still hashes (sha256 of zero bytes). */
export function bodySha256Hex(body: string | Uint8Array): string {
  return createHash("sha256").update(body ?? "").digest("hex");
}

/** The canonical string the HMAC is computed over. */
export function canonicalString(
  method: string,
  path: string,
  timestamp: string,
  nonce: string,
  body: string | Uint8Array,
): string {
  return (
    (method ?? "").toUpperCase() +
    "\n" +
    (path ?? "") +
    "\n" +
    (timestamp ?? "") +
    "\n" +
    (nonce ?? "") +
    "\n" +
    bodySha256Hex(body)
  );
}

/** hex(hmacSha256(apiKey, canonical)) — lowercase hex. */
export function sign(
  apiKey: string,
  method: string,
  path: string,
  timestamp: string,
  nonce: string,
  body: string | Uint8Array,
): string {
  return createHmac("sha256", apiKey ?? "")
    .update(canonicalString(method, path, timestamp, nonce, body))
    .digest("hex");
}

/**
 * Build the three signature headers for a write request. `timestamp` is epoch
 * MILLISECONDS as a string; `nonce` is a fresh UUID. Injectable clock/nonce so
 * callers (and tests) can pin them.
 */
export function signatureHeaders(
  apiKey: string,
  method: string,
  path: string,
  body: string | Uint8Array,
  now: () => number = Date.now,
  nonceGen: () => string = randomUUID,
): Record<string, string> {
  const timestamp = String(now());
  const nonce = nonceGen();
  return {
    [SIGNATURE_HEADERS.timestamp]: timestamp,
    [SIGNATURE_HEADERS.nonce]: nonce,
    [SIGNATURE_HEADERS.signature]: sign(apiKey, method, path, timestamp, nonce, body),
  };
}

File: sdk/go/client/signing.go
package client

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"net/http"
	"strings"
)

// SDK 侧请求签名（方案 A，以 api_key 为 HMAC 密钥）。
//
// 必须与服务端 io.github.pinpols.batch.common.security.RequestSignatures 以及
// Java SDK 的 RequestSigner 逐字节一致 —— 三者由各语言 SDK 的 conformance 用例
// 钉死同一 golden 向量。算法：
//
//	canonical = UPPER(method) "\n" path "\n" timestamp "\n" nonce "\n" hex(sha256(body))
//	signature = hex(hmacSha256(apiKey, canonical))     // 小写 hex
//
// timestamp 为 epoch 毫秒字符串；nonce 建议 UUID；body 为请求原始字节（空 body 也要
// 算 sha256）。
const (
	// HeaderSignatureTimestamp carries the epoch-millis timestamp.
	HeaderSignatureTimestamp = "X-Batch-Timestamp"
	// HeaderSignatureNonce carries the per-request nonce (UUID).
	HeaderSignatureNonce = "X-Batch-Nonce"
	// HeaderSignature carries the lowercase-hex HMAC-SHA256 signature.
	HeaderSignature = "X-Batch-Signature"
)

// bodySha256Hex returns the lowercase-hex SHA-256 of the request body. A nil
// body hashes the empty byte slice (the platform always hashes a body, even an
// absent one).
func bodySha256Hex(body []byte) string {
	sum := sha256.Sum256(body)
	return hex.EncodeToString(sum[:])
}

// canonicalString assembles the canonical signing string. method is uppercased;
// the remaining components are joined verbatim by '\n', with the SHA-256 body
// digest as the final line.
func canonicalString(method, path, timestamp, nonce string, body []byte) string {
	return strings.ToUpper(method) +
		"\n" + path +
		"\n" + timestamp +
		"\n" + nonce +
		"\n" + bodySha256Hex(body)
}

// signRequest computes the lowercase-hex HMAC-SHA256 of the canonical string
// keyed by apiKey. This is the value sent in the X-Batch-Signature header.
func signRequest(apiKey, method, path, timestamp, nonce string, body []byte) string {
	mac := hmac.New(sha256.New, []byte(apiKey))
	mac.Write([]byte(canonicalString(method, path, timestamp, nonce, body)))
	return hex.EncodeToString(mac.Sum(nil))
}

// isWriteMethod reports whether the HTTP method is a mutating (write) request
// (POST/PUT/PATCH/DELETE). Only write requests carry a signature.
func isWriteMethod(method string) bool {
	switch strings.ToUpper(method) {
	case http.MethodPost, http.MethodPut, http.MethodPatch, http.MethodDelete:
		return true
	default:
		return false
	}
}

// newNonce mints a per-request RFC-4122-ish v4 UUID for the X-Batch-Nonce
// header. Only uniqueness/shape matters (the platform treats it as opaque).
func newNonce() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}
